### PR TITLE
[NodeKiller] Rebooting broken nodes before exiting the test

### PR DESCRIPTION
**Pull request by mcastellin**
With this change I tried to introduce some sort of wait mechanism for the NodeKiller subroutine.
The rationale behind is it avoiding that cluster nodes are left in a broken state by NodeKiller if the test finishes while downtime is being simulated for some nodes.

Now the chaos monkey can return back to its caller a WaitGroup that can be used to wait for the subroutine termination before proceeding to test summary.

/assign @jprzychodzen 
Ref: #1130

### CODE REVIEW

**Comment by mcastellin** [comment]
As discussed with @jprzychodzen in #1130 it would also be nice to have a way to mock the NodeKiller so we can implement unit testing for this part of the code. I have submitted the code change for now to get some feedback on the idea. I will try implement some unit testing later to complement this one.

File: clusterloader2/pkg/test/simple_test_executor.go [hunk 1]
```diff
@@ -55,17 +56,50 @@ func (ste *simpleTestExecutor) ExecuteTest(ctx Context, conf *api.Config) *error
...
+				continue
+			}
+		}
+	}
+	klog.Infof(ctx.GetChaosMonkey().Summary())
+	return errList
+}
+
+// ExecuteTestSteps executes all test steps provided in configuration
+func (ste *simpleTestExecutor) ExecuteTestSteps(ctx Context, conf *api.Config) (*sync.WaitGroup, *errors.ErrorList) {
```
Comment (wojtek-t) [inline clusterloader2/pkg/test/simple_test_executor.go]: I like the overall approach, but that it is introducing unnecessary complication.

Let's:
- create stopCh and ChaosMonkey outside of this function (in the ExecuteTest one)
- this one then doesn't need to know (and return) about WaitGroup

Comment (mcastellin) [reply clusterloader2/pkg/test/simple_test_executor.go]: No problem, I refactored it to only return the errorList

**Comment by mcastellin** [comment]
@wojtek-t Thanks for reviewing my work. 
Can you help me out understanding what is wrong with the build? I really can't point my finger on it..

**Comment by jprzychodzen** [comment]
> Can you help me out understanding what is wrong with the build? I really can't point my finger on it..

@mcastellin, what do you mean by 'what is wrong'? It seems that commit passed tests, so it builds. Are you referring to @wojtek-t comments?

**Comment by mcastellin** [comment]
No @jprzychodzen , I was referring to the errors in the first build. The second one passed though.. so I believe it was just a flake.

**Comment by wojtek-t** [comment]
Looks reasonable to me - I will let @jprzychodzen to make a deeper review though.

/approve

**Comment by mcastellin** [comment]
@jprzychodzen 
In addition to this change, I have refactored the NodeKiller to provide two providers:
- A "CloudKill" provider when running in GCE
- a "DummyKill" provider that for now just simulates the chaos monkey with a sleep command

I believe this is what you asked in #1130, right?

If you're interested I can make this part of this PR or follow up with another PR once we decide what to do with this one. What do you think?

**Comment by wojtek-t** [comment]
> If you're interested I can make this part of this PR or follow up with another PR once we decide what to do with this one. What do you think?

Separate PR please. Whenever it's possible, split things into multiple PRs - it makes reviewing easier and faster. Review time is exponential to the size of the PR usually...

**Comment by mcastellin** [comment]
> Separate PR please. Whenever it's possible, split things into multiple PRs - it makes reviewing easier and faster. Review time is exponential to the size of the PR usually...

Agreed, I also though to hold this off. I'll wait until this one is resolved then. Thanks for your feedback @wojtek-t

**Comment by jprzychodzen** [comment]
@mcastellin, I've checked the code, thanks for this contribution!

I need to confirm that this work with few edge cases - there are no unit tests on Node Killer/Chaos Monkey level and right now Chaos Monkey is not enabled in presubmit. I will finish that tomorrow. Some (probably) unrelated problem happened (resource cleanup) and I need to double check that.

Please don't add those provides right now. It would be even better if you could add each provider by a separate PR - smaller PRs are easier to review.

Also, could you squash those three commits into single commit? Thanks!

**Comment by jprzychodzen** [comment]
It seems that the issue I've seen is related to some other problem. I will prepare a replication script and create a separate issue for that.

@mcastellin thanks for contribution.
